### PR TITLE
feat: create new PATCH endpoint for updating form settings

### LINKED ISSUE #1355: Settings API refactor
Related to #815, shard out for priority

### PR DESCRIPTION
**Pull request by karrui**
Ready for review

## Problem

This PR extracts out updating of form settings into its own endpoint `PATCH /form/:formId/settings`.
In addition, this PR lays the groundwork for our API refactor with the `/api/v3` API base url.

Closes #1355, related to #1210

## Solution

**Features**:

- create new ApiRouter accessed via '/api/v3'
- add backend AdminFormService#updateFormSettings fn
- add settings route handler with Joi validation
- add client AdminFormService for form settings update
- add new FormModel instance method `getSettings`

**Improvements**:

- use new updateFormSettings fn to update settings in client

## Tests

Both unit and integration tests have been added for the new PATCH endpoint.

**Manual tests**
- [ ] Update form settings for an **email** mode form. All possible updates should work:
  - [ ] Activate/deactivate form
  - [ ] Toggle submission limit, change limit number
  - [ ] Set inactive form message
  - [ ] Authentication types,  esrvcId, etc.
  - [ ] Toggle on/off captcha
  - [ ] Update form name
  - [ ] Update emails
- [ ] Update form settings for a **storage** mode form. All possible updates should work:
  - [ ] Activate/deactivate form
  - [ ] Toggle submission limit, change limit number
  - [ ] Set inactive form message
  - [ ] Authentication types,  esrvcId, etc.
  - [ ] Toggle on/off captcha
  - [ ] Update form name
  - [ ] Update webhook URL
- [ ] For any mode form, try with random keys to update settings. Should return 400 error

### CODE REVIEW

File: src/app/modules/form/admin-form/admin-form.types.ts
Comment (liangyuanruo) [inline src/app/modules/form/admin-form/admin-form.types.ts:89]: can we rename this to `SettingsUpdateDTO` to emphasise that it is a _data transfer object_?

also, would it be clearer to define it as `Partial<Pick<IForm, ...>>` instead to emphasise that it is a subset of properties from `IForm`?

File: src/public/services/AdminFormService.ts
Comment (liangyuanruo) [inline src/public/services/AdminFormService.ts:3]: I'm thinking we could start numbering the APIs from v3 instead for consistency, as the submission endpoints are already on v2.

Comment (liangyuanruo) [reply src/public/services/AdminFormService.ts:3]: we should also set up a hierarchy of routers for each `/`

File: src/app/modules/form/admin-form/admin-form.service.ts
Comment (liangyuanruo) [inline src/app/modules/form/admin-form/admin-form.service.ts:518]: can we resolve types at the edge of the application (the controller) instead of doing this sort of thing?

File: src/app/modules/form/admin-form/admin-form.service.ts
Comment (liangyuanruo) [inline src/app/modules/form/admin-form/admin-form.service.ts:561]: We should return the Settings object as a single resource, so that the client always has the most updated representation.

File: src/app/modules/form/admin-form/admin-form.service.ts
Comment (karrui) [reply src/app/modules/form/admin-form/admin-form.service.ts:518]: This is not really casting right, it's guaranteed to be type (keyof SettingsUpdateBody), but just telling TypeScript as TypeScript is inferencing it as `string[]` instead.

File: src/app/modules/form/admin-form/admin-form.service.ts
Comment (karrui) [reply src/app/modules/form/admin-form/admin-form.service.ts:561]: I see, so whole object picked by SettingsKeys instead of just the changed settings

File: src/app/modules/form/admin-form/admin-form.types.ts
Comment (karrui) [reply src/app/modules/form/admin-form/admin-form.types.ts]: Resolved, add new `FormSettings` type in `types/form` and `SettingsUpdateDto` type in `shared/typings` in d3afefa

File: src/app/modules/form/admin-form/admin-form.service.ts
Comment (karrui) [reply src/app/modules/form/admin-form/admin-form.service.ts]: resolved in 4276e75

File: src/public/services/AdminFormService.ts
Comment (karrui) [reply src/public/services/AdminFormService.ts]: Ya, this is client side, on backend there is 

ApiRouter `/api/v3` -> AdminRouter `/admin` -> individual endpoints like update form settings `/form/settings`

Comment (karrui) [reply src/public/services/AdminFormService.ts]: Maybe I'll nest another AdminFormRouter for the `/form` endpoint?

Comment (liangyuanruo) [reply src/public/services/AdminFormService.ts]: No, I think you should do it to reflect the hierarchy of the routes. Many projects even have their folder structure organised that way.

File: src/app/modules/form/admin-form/admin-form.service.ts
Comment (liangyuanruo) [inline src/app/modules/form/admin-form/admin-form.service.ts:541]: TypeScript wouldn't compile if the keys were not exhaustive, right?
```suggestion
```

File: src/public/modules/forms/admin/controllers/admin-form.client.controller.js
Comment (liangyuanruo) [inline src/public/modules/forms/admin/controllers/admin-form.client.controller.js:206]: Would this also work?
```suggestion
 Object.assign($scope.myform, updatedSettings)
```

File: src/types/form.ts
Comment (liangyuanruo) [inline src/types/form.ts:151]: should we "inherit" from `IForm` instead, since we only want a subset of a POJO definition?

File: src/public/modules/forms/admin/controllers/admin-form.client.controller.js
Comment (karrui) [reply src/public/modules/forms/admin/controllers/admin-form.client.controller.js:206]: It should, since we are just returning the full subset of settings

File: src/types/form.ts
Comment (karrui) [reply src/types/form.ts:151]: Problem is (maybe not really a problem) that `IForm` has a bunch of optional fields due to them being default in the form schema. The document type has the most correct shape in that all the default fields are no longer optional (as they should already be created as it is now a document)

File: src/app/modules/form/admin-form/admin-form.service.ts
Comment (karrui) [reply src/app/modules/form/admin-form/admin-form.service.ts:541]: Typescript only complains if we specify a return type and there is no default statement, meaning the return type has a possibility to be void.

However, in this case we are mutating an object and may not return anything. Maybe I can return something in the loop instead!

```ts
updateKeys.reduce((accumulatedForm, key) => {
        switch (key) {
          // Non-object patches, assign key to form.
          case 'authType':
          case 'emails':
          case 'esrvcId':
          case 'hasCaptcha':
          case 'inactiveMessage':
          case 'status':
          case 'submissionLimit':
          case 'title': {
            set(accumulatedForm, key, body[key])
            return accumulatedForm
          }
          // Object patches, deep mutate.
          case 'webhook': {
            merge(accumulatedForm[key], body[key])
            return accumulatedForm
          }
        }
```

Now we don't need the whole throwable thing, since yea typescript won't compile! Nice catch, didn't see it then!
image

File: src/app/modules/form/admin-form/admin-form.routes.ts
Comment (karrui) [reply src/app/modules/form/admin-form/admin-form.routes.ts:31]: permissionList is a collaborator thing, not settings right

File: src/shared/typings/core.ts
Comment (karrui) [reply src/shared/typings/core.ts:1]: Good point, got used to the `shared` folder meaning. Should we not move the types to shared instead?

File: src/app/modules/form/admin-form/admin-form.service.ts
Comment (karrui) [reply src/app/modules/form/admin-form/admin-form.service.ts]: Fixed in 5e6be06

File: src/shared/typings/core.ts
Comment (mantariksh) [reply src/shared/typings/core.ts:1]: can also, no difference. but should probably do in a separate PR because nearly every file in the backend will have to update imports haha

File: src/app/modules/form/admin-form/admin-form.routes.ts
Comment (mantariksh) [reply src/app/modules/form/admin-form/admin-form.routes.ts:31]: hmm I think they are separate in the UI, but semantically I'd say the permissionList is part of the settings, right? the alternative is to create another `/collaborators` endpoint which will look nearly identical to this one

Comment (karrui) [reply src/app/modules/form/admin-form/admin-form.routes.ts:31]: Should we not? Even though it may be nearly identical, at least that endpoint only does a single thing and gates properly? Can have PUT collaborators and GET collaborators hahaha

File: src/public/modules/forms/admin/controllers/admin-form.client.controller.js
Comment (karrui) [reply src/public/modules/forms/admin/controllers/admin-form.client.controller.js]: I reused the error message for `400` status. Fixed in 8e63630.

File: src/shared/typings/core.ts
Comment (karrui) [reply src/shared/typings/core.ts:1]: Maybe don't fix it here still, since types are more for backend, and shared is more for both frontend and backend it seems.

File: src/app/modules/form/admin-form/admin-form.controller.ts
Comment (karrui) [reply src/app/modules/form/admin-form/admin-form.controller.ts]: CHeers, missed it. Fixed.

File: src/app/models/form.server.model.ts
Comment (karrui) [reply src/app/models/form.server.model.ts:452]: Done

File: src/public/services/AdminFormService.ts
Comment (karrui) [reply src/public/services/AdminFormService.ts]: Done in 00caf9f

File: src/public/modules/forms/admin/controllers/admin-form.client.controller.js
Comment (karrui) [reply src/public/modules/forms/admin/controllers/admin-form.client.controller.js]: Done in 442e379

File: src/app/modules/form/admin-form/admin-form.routes.ts
Comment (karrui) [reply src/app/modules/form/admin-form/admin-form.routes.ts]: As discussed during syncup, decision is to keep `/collaborators` as a separate endpoint

File: src/shared/typings/core.ts
Comment (mantariksh) [reply src/shared/typings/core.ts:1]: sure, but in that case can we use `types` instead of `typings` for consistency?

Comment (karrui) [reply src/shared/typings/core.ts:1]: I'll move it into `types/api` instead, so we can use types across both frontend and backend as if it was shared instead of fragmenting the places where they exist

Comment (karrui) [reply src/shared/typings/core.ts]: Done in  dc95555

File: src/app/modules/form/admin-form/admin-form.service.ts
Comment (mantariksh) [inline src/app/modules/form/admin-form/admin-form.service.ts:536]: should this just be `formId`, since the id will never change? just for consistency with the rest of the logs

File: src/app/modules/form/form.utils.ts
Comment (mantariksh) [inline src/app/modules/form/form.utils.ts:74]: there's actually already an `isEmailModeForm` function in `src/types/form`

File: src/app/routes/admin/form/__tests__/admin-form.routes.spec.ts
Comment (mantariksh) [inline src/app/routes/admin/form/__tests__/admin-form.routes.spec.ts:37]: dyou wanna put the full path here?

File: src/app/routes/admin/form/__tests__/admin-form.routes.spec.ts
Comment (mantariksh) [inline src/app/routes/admin/form/__tests__/admin-form.routes.spec.ts:55]: why stringify then immediately parse?

**Review by mantariksh — APPROVED** [review]
minor comments, otherwise lgtm!

File: src/app/modules/form/admin-form/admin-form.controller.ts
Comment (liangyuanruo) [inline src/app/modules/form/admin-form/admin-form.controller.ts:996]: I think the API design can be better modeled as an instance method instead i.e.  form.updateSettings(...)? Or perhaps even a static method in Form.updateSettings(...) that relies on Form.findOneAndUpdate() under the hood, so that there won't be concurrency issues from interleaved read-then-write operations.

**Review by liangyuanruo — CHANGES_REQUESTED** [review]

File: src/app/modules/form/form.utils.ts
Comment (karrui) [reply src/app/modules/form/form.utils.ts:74]: As discussed, this was moved from types to utils, since it is a function and not a type.

File: src/app/routes/admin/form/__tests__/admin-form.routes.spec.ts
Comment (karrui) [reply src/app/routes/admin/form/__tests__/admin-form.routes.spec.ts:55]: Converting mongoose object to what the endpoint will actually send out, mimicking real behaviour.

For example, JSON.parse(JSON.stringify()) will remove all keys with `undefined` values (since `undefined` is not representable in the JSON spec), or convert artifacts such as `CoreMongooseArray` to `Array` types for jest to match.

File: src/app/modules/form/admin-form/admin-form.controller.ts
Comment (karrui) [reply src/app/modules/form/admin-form/admin-form.controller.ts:996]: Good point, will make the change.

File: src/types/form.ts
Comment (liangyuanruo) [reply src/types/form.ts:151]: OK. I'll leave it up to you to decide whether it would help to refactor the types or not.

File: src/app/modules/form/admin-form/admin-form.controller.ts
Comment (karrui) [reply src/app/modules/form/admin-form/admin-form.controller.ts:996]: Updated, but note that calling mongoose's `findByIdAndUpdate` requires the update to be in dot notation, or the same issue will occur where objects are overridden instead of merged

File: src/app/modules/form/admin-form/admin-form.service.ts
Comment (liangyuanruo) [inline src/app/modules/form/admin-form/admin-form.service.ts:534]: is this actually possible, since the  `emails` property in are discriminated within `EmailFormSchema`?

File: src/app/modules/form/form.service.ts
Comment (liangyuanruo) [inline src/app/modules/form/form.service.ts:168]: Principle of Least Power - just accept `responseMode` as a parameter?

File: src/app/modules/form/admin-form/admin-form.service.ts
Comment (karrui) [reply src/app/modules/form/admin-form/admin-form.service.ts:534]: It will just pass silently without setting the emails. Do we not want to respond with an error code in that case?

File: src/app/modules/form/form.service.ts
Comment (karrui) [reply src/app/modules/form/form.service.ts:168]: Good point, will fix

File: src/app/routes/api.routes.ts
Comment (liangyuanruo) [inline src/app/routes/api.routes.ts:9]: Would it be simpler to just do this? The `AdminRouter` doesn't need to know where it's mounted, that's decided higher up in the hierarchy in the `ApiRouter`.

```typescript
ApiRouter.use('/admin', AdminRouter)
```

Comment (liangyuanruo) [reply src/app/routes/api.routes.ts:9]: Then we can cut out variables such as `ADMIN_NESTED_ROUTE`

**Review by liangyuanruo — APPROVED** [review]
Pending a couple of small questions, but LGTM otherwise!

File: src/app/modules/form/admin-form/admin-form.service.ts
Comment (liangyuanruo) [reply src/app/modules/form/admin-form/admin-form.service.ts:534]: I think if you wanted to prevent this, then it should be performed at the Joi validator (that's why I suggested always including `responseMode` in the DTO to differentiate). Silent failure isn't great, but I think there's also no harm either.

Comment (liangyuanruo) [reply src/app/modules/form/admin-form/admin-form.service.ts:534]: The real question is are there actually two DTO types, to distinguish between encrypt and email?

File: src/app/routes/api.routes.ts
Comment (karrui) [reply src/app/routes/api.routes.ts:9]: Makes sense. Should we still give it an `ADMIN_ROOT_ROUTE` constant then? Or only do that when we need to make it a constant?

File: src/app/modules/form/admin-form/admin-form.service.ts
Comment (karrui) [reply src/app/modules/form/admin-form/admin-form.service.ts:534]: I'll let it pass silently then. Not really a failure since you shouldn't be giving emails to an encrypt form :P

File: src/app/routes/api.routes.ts
Comment (karrui) [reply src/app/routes/api.routes.ts:9]: I understand now, the route should not be exposed by the router but set by the caller. Will remove all exported route constants and set them in the router

**Pull request merged**
